### DISCUSSION TITLE
Why inference is different every time?

### DISCUSSION THREAD

**Discussion by bzp83** [rhasspy/piper#438] [General]
Every time I run inference with the same input, the output is always different... some times the result is good and other times it's not. It seems like it uses some sort of random number to start it, is this the case? Is this possible to force it to use the same values so the output will always be the same given the same input?

thanks!

**Comment by SeymourNickelson** [comment]
I have the same question. I'm training a model right now and am only 500 epochs in. Feeding the same input multiple times produces slightly different results. I wonder if it's b/c the model hasn't been trained long enough.

**Comment by SeymourNickelson** [comment]
From https://huggingface.co/docs/transformers/main/en/model_doc/vits I found the following info:

"The model is trained end-to-end with a combination of losses derived from variational lower bound and adversarial training. To improve the expressiveness of the model, normalizing flows are applied to the conditional prior distribution. During inference, the text encodings are up-sampled based on the duration prediction module, and then mapped into the waveform using a cascade of the flow module and HiFi-GAN decoder. Due to the stochastic nature of the duration predictor, the model is non-deterministic, and thus requires a fixed seed to generate the same speech waveform."

Just before triggering the infer.py script in piper's repo to generate three samples with the same input (with my most recent checkpoint), I called torch.manual_seed(1234) and got three identical results. Without the torch.manual_seed(1234) the results with the same input are different.

**Reply by bzp83** [reply]
thank you!

**Comment by synesthesiam** [comment]
Noise is added in two places during inference using the `--noise-scale` and `--noise-w` options. As @SeymourNickelson mentioned, you can control this noise by manually seeding torch's random number generator. I would recommend doing that rather than setting the noise to 0, which sounds bad.

**Comment by SeymourNickelson** [comment]
@synesthesiam On my onnx export I do seem to be getting different results with the same input. I see the export onnx script does call torch.manual_seed(1234) before exporting. Do you know what could be causing this?
